File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_so_far = float('inf')
        max_profit = 0
        
        for i in prices:
            min_so_far = min(min_so_far, i)
            max_profit = max(max_profit, i - min_so_far)
            
        return max_profit

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int minsofar = Integer.MAX_VALUE, maxsofar = Integer.MIN_VALUE;
        for(int i=0; i<prices.length; i++) {
            minsofar = minsofar > prices[i] ? prices[i] : minsofar;
            maxsofar = prices[i] - minsofar > maxsofar ? prices[i] - minsofar: maxsofar;
        }
        return maxsofar;
    }
}